cachito/workers/pkg_managers/gomod.py

```diff
@@ -164,8 +168,10 @@ def resolve_gomod(app_source_path, request, dep_replacements=None):
             # This should never occur, but it's here as a precaution
             raise CachitoError(go_module_name_error)
 
+        # NOTE: If there are multiple git modules in a single git repo, they will
+        #   all be versioned identically.
         module_version = get_golang_version(
-            module_name, app_source_path, request["ref"], update_tags=True
+            module_name, git_dir_path, request["ref"], update_tags=True
```

Comment:
No idea how this should behave if the module is in a git submodule. It might be fine to leave that for later, after we figure out how gomod handles these cases. Unless @athos-ribeiro or @lcarva knows?

Reply:
Not that I don't think that this is an important consideration, but I feel like it might warrant a follow-on issue if it is a problem.

Reply:
Does this mean we would be overriding the actual versions of those modules? While this does not seem to be an issue for reproducibility (given the bundle will contain the sources), this would be an issue for grading later, when checking the image content manifests.

One way to go solve the issue would be (to require users) to tag the repository with tags pointing to the submodules (see https://github.com/go-modules-by-example/index/blob/master/009_submodules/README.md).

This would still require additional work on how we process versions and perhaps would require us to check tags in the original git submodule as well to make sure they match.

With that in mind, would it be an issue if, for this first change, we disallow requests with multiple gomod packages AND git submodules to say the operation is not supported?

Reply:
Oh, I didn't realize you can apply git tags to specific subpaths. This breaks even the regular workflow without git submodules, see https://github.com/go-modules-by-example/submodules
```shell
curl "localhost:8080/api/v1/requests" \
    -X POST \
    -H "content-type: application/json" \
    -d '{
          "repo": "https://github.com/go-modules-by-example/submodules",
          "ref": "56ad34e87f3359a8dd4c781941829322edcf0ad6",
          "pkg_managers": ["gomod"],
          "packages": {
            "gomod": [
              {"path": "a"},
              {"path": "b"}
            ]
          }
        }'
```
```
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/celery/app/trace.py", line 385, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/usr/lib/python3.8/site-packages/celery/app/trace.py", line 648, in __protected_call__
    return self.run(*args, **kwargs)
  File "/src/cachito/workers/tasks/gomod.py", line 97, in fetch_gomod_source
    gomod = resolve_gomod(
  File "/src/cachito/workers/pkg_managers/gomod.py", line 173, in resolve_gomod
    module_version = get_golang_version(
  File "/src/cachito/workers/pkg_managers/gomod.py", line 492, in get_golang_version
    pseudo_base_tag = _get_highest_semver_tag(repo, commit, major_version, all_reachable=True)
  File "/src/cachito/workers/pkg_managers/gomod.py", line 433, in _get_highest_semver_tag
    return repo.tags[highest]
  File "/usr/lib/python3.8/site-packages/git/util.py", line 898, in __getitem__
    raise IndexError("No item found with id %r" % (self._prefix + index))
IndexError: No item found with id 'v1.0.0'
```

Reply:
> With that in mind, would it be an issue if, for this first change, we disallow requests with multiple gomod packages AND git submodules to say the operation is not supported?

Yes, it would break building https://github.com/openshift/windows-machine-config-operator

Reply:
We could fail requests that use both gomod and git-submodule *only* if they use non-root gomod packages, like this: https://github.com/release-engineering/cachito/pull/333#discussion_r518775196
So builds with configuration like this would still work:
```json
{
  "pkg_managers": ["gomod", "git-submodule"]
}
```
What would *not* work is this:
```json
{
  "pkg_managers": ["gomod", "git-submodule"],
  "packages": {
    "gomod": [{"path": "foo"}]
  }
}
```
"gomod" in the "packages" dict was not allowed until this PR, so this will not break existing builds.

Reply:
This comment is outdated since I modified the `NOTE` in the code. It now reads

```
# NOTE: If there are multiple go modules in a single git repo, they will
#   all be versioned identically.
```

There are two different possible issues that are presented in this thread:

1. How to handle the versioning of multiple modules within the same git repo
2. How to handle the versioning of modules across git submodules

Is the proper resolution for both of these the same? I feel like 1 seems like a more non-blocking issue than 2 does.

@chmeliik, That seems like a reasonable approach to me. It would allow people to use each of the features independently, but explicitly disallow their use together. As I commented in https://github.com/release-engineering/cachito/pull/333#discussion_r518775196, the issue is really when the paths collide, but that approach could prevent us from getting into some untested/unsupported areas in the short term.

Reply:
The resolutions might be different:
1. Handle git tags which apply to a specific subpath (e.g. `module_a/v1.0.0`)
2. [?] Take git tags from the submodule, not the main repo (but also guard against inconsistency with 1)

For 2, I think we still need to verify if gomod does this

Agreed that 1 seems less important for now, and 2 will be blocked as we discussed.
Will file stories to follow up on both, for now this seems like a good approach